src/sdi-progress-dock.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2020-2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "sdi-progress-dock.h"
+#include "com.canonical.Unity.LauncherEntry.h"
+#include <glib/gi18n.h>
+#include <snapd-glib/snapd-glib.h>
+
+/**
+ * This class manages the progress bars in the dock for each snap
+ * being updated.
+ */
+
+struct _SdiProgressDock {
+  GObject parent_instance;
+
+  GApplication *application;
+  UnityComCanonicalUnityLauncherEntry *unity_manager;
+};
+
+G_DEFINE_TYPE(SdiProgressDock, sdi_progress_dock, G_TYPE_OBJECT)
+
+/**
+ * This callback should be connected to the `refresh-progress` signal from a
+ * #sdi_refresh_monitor object. It will receive the total number of tasks and
+ * how many have been done, and if the task has been completed, and with that
+ * will update the progress bars in the dock.
+ */
+void sdi_progress_dock_update_progress(SdiProgressDock *self, gchar *snap_name,
+                                       GStrv desktop_files,
+                                       gchar *task_description,
+                                       guint done_tasks, guint total_tasks,
+                                       gboolean task_done) {
+  if (desktop_files != NULL && total_tasks != 0) {
```

Comment:
Nit, just return early in this case:

```suggestion
  if (desktop_files == NULL || !total_tasks)
     return;
```

Reply:
Done, although being `total_tasks` a counter, I prefer to keep the comparison with zero.

Reply:
Yeah, fair enough... I was undecided too :)